goldenmaster/apigear/mqtt/mqttcppclient.cpp

```diff
@@ -0,0 +1,578 @@
+        MQTTAsync_destroy(m_client.get());
+        m_client.reset();
+
+        // reconnect if the connection was not dropped intentionally by us
+        if(!disconnectRequested)
+        {
+            // we need to re-subscribe to all topics on re-connection
+            m_subscribedTopicsMutex.lock();
+            m_toBeSubscribedTopicsMutex.lock();
+            if(!m_subscribedTopics.empty())
+            {
+                m_subscribedTopics.insert(m_toBeSubscribedTopics.begin(),m_toBeSubscribedTopics.end());
+                m_toBeSubscribedTopics = m_subscribedTopics;
+                m_subscribedTopics.clear();
+            }
+            m_toBeSubscribedTopicsMutex.unlock();
+            m_subscribedTopicsMutex.unlock();
+
+            connectToHost("");
+        }
+    }).detach();
+}
+
+void Client::handleTextMessage(const Message& message)
+{
+    // std::clog << "new msg topic: " << topic << " mqtt_message: " << message << std::endl;
+    std::pair<std::multimap<Topic, ISink*, Topic>::iterator, std::multimap<Topic, ISink*, Topic>::iterator> topics = m_subscribedTopics.equal_range(message.topic);
+    // for (auto test: m_subscribedTopics){
+    //     std::cout << "topic " << test.first << " " << test.second << std::endl; 
+    // }
+    for (auto iter = topics.first; iter != topics.second; ++iter)
+    {
+        if(iter->second != nullptr)
+        {
+            if(message.topic.getTopicType() == Topic::TopicType::Signal)
+            {
+                iter->second->onSignal(message.topic, message.content);
+            }
+            else if(message.topic.getTopicType() == Topic::TopicType::Property)
+            {
+                iter->second->onPropertyChanged(message.topic, message.content);
+            }
+            else if(message.topic.getTopicType() == Topic::TopicType::Operation)
+            {
+                iter->second->onInvoke(message.topic, message.content, message.responseTopic, message.correlationData);
+            }
+        }
+        else
+        {
+            // call locally
+            const int randomId = std::stoi(message.correlationData);
+            InvokeReplyFunc func {};
+            m_invokeResultsMutex.lock();
+            if((m_invokeResults.find(randomId) != m_invokeResults.end()))
+            {
+                func = m_invokeResults[randomId];
+                m_invokeResults.erase(randomId);
+            }
+            m_invokeResultsMutex.unlock();
+            if(func) {
+                const InvokeReplyArg arg{nlohmann::json::parse(message.content)};
+                func(arg);
+            }
+        }
+    }
+}
+
+void Client::invokeRemote(const Topic& topic, const std::string& value, InvokeReplyFunc func)
+{
+    const std::string responseTopic = topic.getEncodedTopic() + "/" + m_clientID + "/result" ;
+
+    MQTTAsync_responseOptions opts = MQTTAsync_responseOptions_initializer;
+    MQTTAsync_message pubmsg = MQTTAsync_message_initializer;
+    int rc = 0;
```

Comment:
same as for other declare it near the if initilized with the function expression and just compare output in if

Reply:
done